File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode(-1)
        prev = head
        cur1 = list1
        cur2 = list2
        while cur1 and cur2:
            if cur1.val > cur2.val:
                prev.next = cur2
                cur2 = cur2.next
            else:
                prev.next = cur1
                cur1 = cur1.next
            prev = prev.next
        if cur1:
            prev.next = cur1
        elif cur2:
            prev.next = cur2
        return head.next

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
    if (lists.size() == 0)
      return nullptr;
    for (int i = 1; i < lists.size(); i++) {
      lists[i] = mergeLists(lists[i], lists[i - 1]);
    }
    return lists[lists.size() - 1];
  }
  ListNode* mergeLists(ListNode* list1, ListNode* list2) {
    ListNode* res = new ListNode(-1);
    ListNode* dummy = res;
    while (list1 != nullptr && list2 != nullptr) {
      if (list1->val < list2->val) {
        dummy->next = new ListNode(list1->val);
        list1 = list1->next;
      } else {
        dummy->next = new ListNode(list2->val);
        list2 = list2->next;
      }
      dummy = dummy->next;
    }
    if (list1 != nullptr) {
      dummy->next = list1;
    }
    if (list2 != nullptr) {
      dummy->next = list2;
    }
    return res->next;
  }


};
